models/shopcart.py

```diff
@@ -39,3 +38,36 @@ def all():
     def remove_all():
         """ Remove all Shopcarts from memory """
         Shopcart.__data = []
+
+    @staticmethod
+    def __validate_products(products):
+        """ Validates products or raises an error"""
+        try:
+            # Product is none so set an empty list
+            if products is None:
+                return {}
+            
+            #Not None
+            # Tuple of product, quantity 
+            if type(products) == tuple  and len(products)==2 and all(isinstance(pq,int) for pq in products): 
+                return {products[0]:products[1]}
+
+            # Just a Product id, set default quantity to 1
+            if type(products) == int: 
+                return {products:1}
+
+            if type(products) != dict : 
+                raise DataValidationError("Invalid format for products")
+
+            
+            # Products is a dict of proper tuples
+            if ( all( isinstance(pid,int) for pid in products.keys() ) and
+                 all( (isinstance(q,int) and (q > 0)) for q in products.values() ) ):
+                return products
+
+            #Products not valid
+            raise DataValidationError("Invalid format for products")
+
+        except DataValidationError as e:
```

Comment:
You're still catching the `DataValidationError` and returning an empty products object. We want to unwrap this code from the try-except statement so the error is actually thrown to the caller and not handled within the method.

Reply:
Great! I think it's done now.
@radhikamattoo